exporter/datadogexporter/metadata/host.go

```diff
@@ -0,0 +1,64 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metadata
+
+import (
+	"runtime"
+
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter/config"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter/metadata/system"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter/utils/cache"
+)
+
+// GetHost gets the hostname according to configuration.
+// It gets the configuration hostname and if
+// not available it relies on the OS hostname
+func GetHost(logger *zap.Logger, cfg *config.Config) *string {
+	if cacheVal, ok := cache.Get(cache.CanonicalHostnameKey); ok {
+		return cacheVal.(*string)
+	}
+
+	if err := validHostname(cfg.Hostname); err == nil {
+		cache.SetNoExpire(cache.CanonicalHostnameKey, &cfg.Hostname)
+		return &cfg.Hostname
+	} else if cfg.Hostname != "" {
+		logger.Error("Hostname set in configuration is invalid", zap.Error(err))
+	}
+
+	// Get system hostname
+	hostInfo := system.GetHostInfo(logger)
```

Comment:
It makes sense to me to have a method on `HostInfo` to handle which one to return (I will change that), but I think passing the configured hostname will be a little cumbersome in the future. The end goal is to have other metadata sources (like EC2 or GCE metadata) that are considered in a certain order (a very simplified version of [this code in our Datadog Agent](https://github.com/DataDog/datadog-agent/blob/56f48f501fbef7403736e333ab8365de80906dd1/pkg/util/hostname.go#L132-L293)) and I would prefer to avoid passing the configured hostname to each of them

Reply:
I reorganized the code so that some of the fallback logic is handled in a `HostInfo` method https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/1b0973a433d13fe7e50c253764d04ff58e5f470a/exporter/datadogexporter/metadata/system/host_unix.go#L42-L50